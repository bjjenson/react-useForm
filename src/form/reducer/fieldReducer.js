import { List, Map } from 'immutable'
import { syncListIndexes } from './syncListIndexes'
import { mergeValidationResults } from './mergeValidationResults'
import { mergeFormValues } from '../helpers/mergeFormValues'

export const actionTypes = {
  insertField: 'insertField',
  removeField: 'removeField',
  addListItem: 'addListItem',
  removeListItem: 'removeListItem',
  updateListIndex: 'updateListIndex',
  reset: 'reset',
  touched: 'touched',
  updateValue: 'updateValue',
  validateAll: 'validateAll',
  validationResult: 'validationResult',
  addListener: 'addListener',
  removeListener: 'removeListener',
}

export const actions = {
  insertField: (fieldName, fieldState) => ({ type: actionTypes.insertField, fieldName, payload: fieldState }),
  removeField: fieldName => ({ type: actionTypes.removeField, fieldName }),
  addListItem: (fieldName, fieldState) => ({ type: actionTypes.addListItem, fieldName, payload: fieldState }),
  removeListItem: (fieldName, index) => ({ type: actionTypes.removeListItem, fieldName, payload: index }),
  updateListIndex: (fieldName, fromIndex, toIndex) => ({ type: actionTypes.updateListIndex, fieldName, payload: { fromIndex, toIndex } }),
  reset: (state) => ({ type: actionTypes.reset, payload: state }),
  touched: (fieldName) => ({ type: actionTypes.touched, fieldName }),
  updateValue: (fieldName, value) => ({ type: actionTypes.updateValue, fieldName, payload: value }),
  validateAll: (errors) => ({ type: actionTypes.validateAll, payload: errors }),
  validationResult: (fieldName, error, helperText) => ({ type: actionTypes.validationResult, fieldName, payload: { error, helperText } }),
  addListener: (fieldName, listener) => ({ type: actionTypes.addListener, fieldName, payload: listener }),
  removeListener: (fieldName, listener) => ({ type: actionTypes.removeListener, fieldName, payload: listener }),
}

export const removedFieldsKey = 'removedFields'
export const fieldsKey = 'fields'
export const listenersKey = 'listeners'

export const initState = state => {
  return state
}

const current = 'current'

export const fieldReducer = (state, { type, fieldName = '', payload }) => {
  const fieldPath = getFieldPath(fieldName)

  const handlers = {
    [actionTypes.updateValue]: value => {
      setTimeout(() => {
        const previous = state.getIn([fieldsKey, ...fieldPath, current, 'value'], '')
        state.getIn([listenersKey, fieldName], List()).forEach(listener => listener(value, previous, state.get('formTools').current))
      })

      const newState = state.setIn([fieldsKey, ...fieldPath, current, 'value'], value)
        .setIn([fieldsKey, ...fieldPath, current, 'pristine'], value == state.getIn([fieldsKey, fieldName, 'initial', 'value']))
      setTimeout(() => {
        const onFormChange = state.get('onFormChange')
        if (onFormChange && typeof (onFormChange) === 'function') {
          onFormChange(mergeFormValues(newState, Map()))
        }
      })

      return newState
    },

    [actionTypes.touched]: () =>
      state.setIn([fieldsKey, ...fieldPath, current, 'touched'], true),

    [actionTypes.validationResult]: ({ error, helperText }) => {
      return state.setIn([fieldsKey, ...fieldPath, current, 'error'], error)
        .setIn([fieldsKey, ...fieldPath, current, 'helperText'],
          helperText || state.getIn([fieldsKey, ...fieldPath, 'initial', 'field']).helperText
        )
    },

    [actionTypes.validateAll]: (errors) =>
      mergeValidationResults(state, errors, true),

    [actionTypes.insertField]: fieldState =>
      state.setIn([fieldsKey, ...fieldPath], fieldState)
        .deleteIn([removedFieldsKey, state.get(removedFieldsKey, List()).indexOf(fieldName)]),

    [actionTypes.removeField]: () =>
      state.deleteIn([fieldsKey, ...fieldPath])
        .set(removedFieldsKey, state.get(removedFieldsKey, List()).push(fieldName)),

    [actionTypes.addListItem]: itemState => {
      const path = [fieldsKey, ...fieldPath, 'items']
      return state.updateIn([...path], List(), prevState => prevState.push(itemState))
    },

    [actionTypes.removeListItem]: index =>
      syncListIndexes(
        state.deleteIn([fieldsKey, ...fieldPath, 'items', index]),
        fieldsKey,
        fieldPath,
      ),

    [actionTypes.updateListIndex]: ({ fromIndex, toIndex }) => {
      const path = [fieldsKey, ...fieldPath, 'items']
      const item = state.getIn([...path, fromIndex])
      return syncListIndexes(state.deleteIn([...path, fromIndex]).updateIn(path, items => items.insert(toIndex, item)), fieldsKey, fieldPath)
    },

    [actionTypes.addListener]: listener =>
      state.updateIn([listenersKey, fieldName], List(), listeners =>
        listeners.push(listener)
      ),

    [actionTypes.removeListener]: listener =>
      state.updateIn([listenersKey, fieldName], List(), listeners => {
        const index = listeners.indexOf(listener)
        return index > -1 ? listeners.delete(index) : listeners
      }),

    [actionTypes.reset]: state => state,
  }

  const handler = handlers[type]
  if (handler) {

    let lastPath = fieldPath
    if (!areArraysEqual(state.get('lastPath', []), lastPath)) {
      lastPath = []
    }

    const trackLastPath = type !== actionTypes.validationResult && type !== actionTypes.reset
    return handler(payload)
      .set('lastPath', trackLastPath ? lastPath : [])
  }
  return state
}

export const getFieldPath = (fieldName, prev = []) => {
  let fieldPath = [fieldName]
  const match = fieldName.match(/.items.(\d)./)

  if (match) {
    const firstField = fieldName.substr(0, match.index)
    const rest = fieldName.match(/.items.\d.fields.(.*)$/)[1]

    if (rest.match(/.items.(\d)./)) {
      return getFieldPath(rest, [firstField, 'items', match[1], 'fields'])
    }

    fieldPath = [...prev, firstField, 'items', match[1], 'fields', rest]

  }

  return fieldPath
}

const areArraysEqual = (a, b) => {
  if (a.length !== b.length) return false

  let same = true
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) { same = false }
  }
  return same
}
