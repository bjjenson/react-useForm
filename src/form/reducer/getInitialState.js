import { fromJS, Map } from 'immutable'
import { fieldsKey } from './fieldReducer'
import { generateDefaultListState } from './generateDefaultListState'
import { generateDefaultFieldState } from './generateDefaultFieldState'

/**
 * @param fields {Array<import("..").IFormFieldArgs>}
 */
export const getInitialState = (fields, initialValues = Map(), options = {}) => {
  const fieldMap = fields.reduce((acc, field) => {
    const result = acc.set(field.name, getFieldState(field, initialValues, options))

    return result
  }, Map())

  return fromJS({ [fieldsKey]: fieldMap })
}

const getFieldState = (field, initialValues, options) => {
  if (field.type === 'list') {
    return generateDefaultListState(field, initialValues, options)
  }
  return generateDefaultFieldState(field, initialValues, options)
}

