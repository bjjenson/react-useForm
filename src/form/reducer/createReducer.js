import { Map } from 'immutable'
import { useRef, useReducer, useMemo } from 'react'
import { fieldReducer, actions } from './fieldReducer'
import { getInitialState } from './getInitialState'

export const createReducer = ({ fields, options = {}, initialValues = Map(), formTools }) => {
  const calcInitialState = () => {
    return getInitialState(fields, initialValues, options, formTools)
  }
  const [state, dispatch] = useReducer(fieldReducer, Map(), () => {
    return calcInitialState()
  })
  const hashCodeRef = useRef()

  useMemo(() => {
    if(!hashCodeRef.current) {
      hashCodeRef.current = initialValues.hashCode()
      return
    }

    hashCodeRef.current = initialValues.hashCode()
    setTimeout(() => {
      dispatch(actions.reset(calcInitialState()))
      options.initialized && options.initialized({ initialValues, ...formTools.current })
    })
  }, [initialValues.hashCode()])

  return [state, dispatch]
}
