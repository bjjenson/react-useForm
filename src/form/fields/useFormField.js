import { actions } from '../reducer/fieldReducer'
/**
 * @param  fieldArgs { import("../useForm").IFormFieldArgs}
 * @returns {import("../useForm").IFormField}
 */
export const useFormField = (state, dispatch, fieldArgs = {}) => {
  const requiredMessage = fieldArgs.requiredMessage || 'Required'

  const tryValidate = (value, touched) => {
    let result
    if (value && fieldArgs.validate) {
      result = fieldArgs.validate(value, fieldArgs.name, state.get('getAllValues'))
    }

    if (!Boolean(result) && !fieldArgs.optional && touched && value === '') {
      result = requiredMessage
    }

    if (touched) {
      dispatch(actions.validationResult(fieldArgs.name, Boolean(result), result))
    }
    return !Boolean(result)
  }

  const setValue = v => {
    dispatch(actions.updateValue(fieldArgs.name, v))
    tryValidate(v, state.getIn(['current', 'touched']))
  }

  const onChange = event => {
    const value = fieldArgs.valueFromChange ? fieldArgs.valueFromChange(event) : event.target.value
    const coercedValue = fieldArgs.normalize ? fieldArgs.normalize(value) : value
    setValue(coercedValue)
    tryValidate(coercedValue, true)
  }

  const onBlur = () => {
    dispatch(actions.touched(fieldArgs.name))
    tryValidate(state.getIn(['current', 'value']), true)
  }

  const setValidationResult = result => {
    dispatch(actions.validationResult(fieldArgs.name, true, result))
  }

  const validate = () => {
    return tryValidate(state.getIn(['current', 'value']), true)
  }

  return {
    props: {
      error: state.getIn(['current', 'error']),
      helperText: state.getIn(['current', 'helperText']),
      label: state.getIn(['initial', 'label']),
      value: state.getIn(['current', 'value']),
      // handlers
      onBlur,
      onChange,
    },
    meta: {
      touched: state.getIn(['current', 'touched']),
      pristine: state.getIn(['current', 'pristine']),
    },
    setValidationResult,
    setValue,
    validate,
  }
}
