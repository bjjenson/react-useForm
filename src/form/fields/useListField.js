import { List, Map } from 'immutable'
import { actions } from '../reducer/fieldReducer'
import { resolveFieldData, getFieldProps } from '../resolveFieldData'
import { getFields, generateDefaultListState } from '../reducer/generateDefaultListState'
import { prepareNameForValidate } from './prepareNameForValidate'
import {getId} from '../helpers/getId'

/**
 * @param  fieldArgs { import("../useForm").IFormFieldArgs}
 * @returns {import("../useForm").IFormField}
 */
export const useListField = (state, dispatch, fieldArgs = {}, getAllValues, formOptions = {}) => {
  const requiredMessage = fieldArgs.requiredMessage || 'Required'

  const setValue = (value) => {
    const s = generateDefaultListState(fieldArgs, Map().setIn(fieldArgs.name.split('.'), value), formOptions)

    dispatch(actions.updateList(fieldArgs.name, s.get('items')))
    validate(s.get('items'))
  }

  const validate = (items) => {
    let message
    if (fieldArgs.validate) {
      message = fieldArgs.validate(items, prepareNameForValidate(fieldArgs.name), state.get(('getAllValues')))
    }
    if (!message && items.size === 0 && !fieldArgs.optional) {
      message = requiredMessage
    }

    dispatch(actions.validationResult(fieldArgs.name, Boolean(message), message || ''))
  }

  const add = (item = Map()) => {
    const options = state.getIn(['initial', 'options'])
    const field = state.getIn(['initial', 'field'])
    const items = state.get('items', List())
    const itemState = getFields(field.fields, item, options, fieldArgs.name, items.size)
    dispatch(actions.addListItem(fieldArgs.name, itemState))
    validate(items.push(item))
  }

  const remove = index => {
    const items = state.get('items', List())
    dispatch(actions.removeListItem(fieldArgs.name, index))
    validate(items.remove(index))
  }

  const updateIndex = (fromIndex, toIndex) => {
    dispatch(actions.updateListIndex(fieldArgs.name, fromIndex, toIndex))
  }

  const fieldData = state.getIn(['items'], List()).map(item => {
    return resolveFieldData(item, dispatch, getAllValues, formOptions)
  }).toArray()

  const id = getId(formOptions.id, fieldArgs.name)
  return {
    props: {
      id,
      error: state.getIn(['current', 'error']),
      helperText: state.getIn(['current', 'helperText']),
      label: state.getIn(['initial', 'label']),
      items: fieldData.map((item, index) => {
        const key = `${fieldArgs.name}.${index}`
        return {
        ...getFieldProps(item, state.getIn(['items', index]), getId(formOptions.id, key)),
        key,
      }}),
      add,
      remove,
      updateIndex,
    },
    setValue,
  }
}

export const defaultListValue = List()
