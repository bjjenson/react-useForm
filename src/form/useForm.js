import { useRef, useEffect } from 'react'
import { Map, fromJS } from 'immutable'
import Form from './Form'
import { createReducer } from './reducer/createReducer'
import { getFieldState } from './reducer/getFieldState'
import { actions, fieldsKey } from './reducer/fieldReducer'
import { getInitialState } from './reducer/getInitialState'
import { resolveFieldData, resolveField, getFieldProps } from './resolveFieldData'
import { mergeFormValues } from './helpers/mergeFormValues'
import { getFieldValues } from './helpers/getFieldValues'
import { formStateResolvers } from './formStateResolvers'
import { setPropsAtPath } from './helpers/setPropsAtPath'
import { validateAll } from './validate/validateAll'
import { getHasError } from './validate/getHasError'
import { pruneNonErrors } from './validate/pruneNonErrors'

/**
 * @param  param0 { import("./useForm").IFormProps }
 */
export const useForm = ({ fields, submit, validate, options = {}, initialValues = Map() }) => {
  const formTools = useRef()
  const [state, dispatch] = createReducer({ fields, initialValues, options, formTools })
  const fieldCache = useRef()

  let fieldData
  const lastPath = state.get('lastPath')

  const getAllValues = () => {
    return fromJS(getFieldValues(fieldCache.current))
  }

  const t0 = performance.now()
  if (!lastPath || lastPath.length === 0 || lastPath[0] === '') {
    fieldData = resolveFieldData(state, dispatch, getAllValues, options)
  } else {
    const path = [fieldsKey, ...lastPath]
    const props = resolveField(state.getIn(path), dispatch, getAllValues)
    fieldData = fieldCache.current
    setPropsAtPath(fieldData, props, lastPath)
  }
  if (options.logPerformance) {
    console.log('Perf:resolveFieldData', (performance.now() - t0))
  }

  fieldCache.current = fieldData

  const stateResolvers = formStateResolvers(state)

  const addField = field => {
    const fieldState = getFieldState(field, initialValues, options)
    dispatch(actions.insertField(field.name, fieldState))
  }

  const removeField = fieldName => {
    dispatch(actions.removeField(fieldName))
  }

  const addFieldListener = (fieldName, listener) => {
    dispatch(actions.addListener(fieldName, listener))
  }

  const removeFieldListener = (fieldName, listener) => {
    dispatch(actions.removeListener(fieldName, listener))
  }

  const getValidationResult = () => {
    let errors = validateAll(state, fieldData, validate, getAllValues)
    dispatch(actions.validateAll(errors))
    errors = pruneNonErrors(errors)
    return {
      isValid: Object.keys(errors).length === 0,
      errors,
    }
  }

  const trySubmitTheForm = (skipValidation = false) => {
    let canSkip = false
    let canSubmit
    if (typeof skipValidation === 'boolean') {
      canSkip = skipValidation
    }

    if (canSkip) {
      canSubmit = true
    } else {
      const errors = validateAll(state, fieldData, validate, getAllValues)
      canSubmit = !getHasError(errors)
      dispatch(actions.validateAll(errors))
    }

    if (canSubmit && submit) {
      submit(mergeFormValues(state, initialValues))
    }
  }

  const getValuesIfFormValid = () => {
    const errors = validateAll(state, fieldData, validate, getAllValues)

    if (!getHasError(errors)) {
      return mergeFormValues(state, initialValues)
    }
    return null
  }

  const resetForm = () => {
    const derivedInitialState = getInitialState(fields, initialValues, options, formTools)
    dispatch(actions.reset(derivedInitialState))
  }

  const setValue = (fieldName, value) => {
    fieldData[fieldName].setValue(value)
  }

  const getValues = () => mergeFormValues(state, initialValues)

  const updateFieldDefinition = (fieldName, definition) => {
    dispatch(actions.updateFieldDefinition(fieldName, { definition: { name: fieldName, ...definition }, options }))
  }

  const fieldProps = getFieldProps(fieldData, state, options.id)

  formTools.current = {
    addField,
    removeField,
    addFieldListener,
    removeFieldListener,
    setValue,
    getValues,
    getFields: () => fieldProps,
    updateFieldDefinition,
  }

  useEffect(() => {
    options.initialized && options.initialized({ initialValues, ...formTools.current })
  }, [initialValues.hashCode()])

  return [
    { ...fieldProps },
    {
      setValue,
      getValuesIfFormValid,
      submit: trySubmitTheForm,
      validate: getValidationResult,
      reset: resetForm,
      id: options.id || '',
      Form,
      addField,
      removeField,
      addFieldListener,
      removeFieldListener,
      ...stateResolvers,
      getValues,
      updateFieldDefinition,
    },
  ]
}

