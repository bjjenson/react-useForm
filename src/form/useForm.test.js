import { fromJS } from 'immutable'
import { useForm } from './useForm'
import { useFormField } from './fields/useFormField'
import { createReducer } from './reducer/createReducer'
import { mergeFormValues } from './helpers/mergeFormValues'
import { getInitialState } from './reducer/getInitialState'

jest.mock('react')
jest.mock('./fields/useFormField')
jest.mock('./reducer/createReducer')
jest.mock('./helpers/mergeFormValues')
jest.mock('./reducer/getInitialState')

const submitWorker = jest.fn()
const dispatch = jest.fn()

const createState = fields => fromJS(fields.reduce((acc, field) => {
  acc.fields[field.name] = {
    initial: {
      type: field.type,
      label: field.label,
    },
    current: {
      value: `current value of ${field.name}`,
    },
  }
  return acc
}, { fields: {} }))

let fields, fieldProps, initialValues

beforeEach(() => {
  fields = [
    { name: 'name', type: 'text', label: 'The name' },
    { name: 'phone', type: 'text', label: 'The phone' },
  ]
  fieldProps = {
    validate: jest.fn(),
    setValidationResult: jest.fn(),
    setValue: jest.fn(),
  }
  initialValues = fromJS({
    nested: {
      name: 'old name',
    },
    id: 'id',
  })

  fieldProps.validate.mockReturnValue(true)

  useFormField.mockImplementation((state) => ({
    ...fieldProps,
    props: {
      label: state.getIn(['initial', 'label']),
      value: state.getIn(['current', 'value']),
    },
  }))
  createReducer.mockImplementation(({ fields }) => [createState(fields), dispatch])
  mergeFormValues.mockReturnValue('merged form values')
  getInitialState.mockReturnValue('derived-initial-state')
})

test('returns array of form props', () => {
  const actual = useForm({ fields })
  expect(actual).toMatchSnapshot()
})

test('no submit if field validation fails', () => {
  fieldProps.validate.mockReturnValue(false)

  const [, { submit }] = useForm({ fields, submit: submitWorker })
  submit()
  expect(submitWorker).not.toHaveBeenCalled()
})

test('no submit if form validation fails', () => {
  const validateForm = jest.fn()
  validateForm.mockReturnValue({ name: 'i am error' })
  const [, { submit }] = useForm({ fields, worker: submitWorker, validate: validateForm })
  submit()
  expect(validateForm.mock.calls[0]).toMatchSnapshot()
  expect(submitWorker).not.toHaveBeenCalled()
})

test('validate can be an array of validators', () => {
  const validate1 = jest.fn()
  const validate2 = jest.fn()
  validate1.mockReturnValue({ name: 'i am name error' })
  validate2.mockReturnValue({ phone: 'i am phone error' })
  const values = { 'name': 'current value of name', 'phone': 'current value of phone' }

  const [, { submit }] = useForm({ fields, worker: submitWorker, validate: [validate1, validate2] })

  submit()
  expect(validate1).toHaveBeenCalledWith(values)
  expect(validate2).toHaveBeenCalledWith(values)
  expect(submitWorker).not.toHaveBeenCalled()
})

test('valid form submits values ', () => {
  fields = [
    { name: 'nested.name' },
    { name: 'phone' },
  ]

  const [, { submit }] = useForm({ fields, submit: submitWorker, initialValues })

  submit()
  expect(submitWorker.mock.calls[0]).toMatchSnapshot()
  expect(mergeFormValues.mock.calls[0]).toMatchSnapshot()
})

test('options passed to createReducer', () => {
  const options = {
    optional: 'yes',
  }
  useForm({ fields, submit: submitWorker, options })

  expect(createReducer.mock.calls[0]).toMatchSnapshot()
})

test('reset dispatches action with derived intial state', () => {
  const [, { reset }] = useForm({ fields, initialValues })

  reset()
  expect(getInitialState.mock.calls[0]).toMatchSnapshot()
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})
