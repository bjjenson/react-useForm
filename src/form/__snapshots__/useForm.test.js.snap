// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addField 1`] = `
Array [
  Object {
    "name": "newField",
    "type": "list",
  },
  Immutable.Map {
    "nested": Immutable.Map {
      "name": "old name",
    },
    "id": "id",
  },
  Object {},
]
`;

exports[`addField 2`] = `
Array [
  Object {
    "fieldName": "newField",
    "payload": "i am a new field",
    "type": "insertField",
  },
]
`;

exports[`addFieldListener dispatches action 1`] = `
Array [
  Object {
    "fieldName": "formField",
    "payload": "myListener",
    "type": "addListener",
  },
]
`;

exports[`dispatches updateFields from tools.updateFieldDefinition  1`] = `
Array [
  Object {
    "fieldName": "field-name",
    "payload": Object {
      "definition": Object {
        "name": "field-name",
        "optional": true,
      },
      "options": Object {},
    },
    "type": "updateFieldDefinition",
  },
]
`;

exports[`form.validate returns all errors 1`] = `
Object {
  "errors": "pruned errors",
  "isValid": false,
}
`;

exports[`form.validate returns all errors 2`] = `
Array [
  Immutable.Map {
    "fields": Immutable.Map {
      "name": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The name",
        },
        "current": Immutable.Map {
          "value": "current value of name",
        },
      },
      "phone": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The phone",
        },
        "current": Immutable.Map {
          "value": "current value of phone",
        },
      },
    },
  },
  Object {
    "name": Object {
      "props": Object {
        "label": "The name",
        "value": "current value of name",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
    "phone": Object {
      "props": Object {
        "label": "The phone",
        "value": "current value of phone",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
  },
  Array [
    "i validate forms",
    "as do i",
  ],
  [Function],
]
`;

exports[`form.validate returns all errors 3`] = `
Array [
  Object {
    "payload": Object {
      "obj": "containing errors",
    },
    "type": "validateAll",
  },
]
`;

exports[`getValuesIfFormValid returns null if not valid 1`] = `
Array [
  Immutable.Map {
    "fields": Immutable.Map {
      "name": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The name",
        },
        "current": Immutable.Map {
          "value": "current value of name",
        },
      },
      "phone": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The phone",
        },
        "current": Immutable.Map {
          "value": "current value of phone",
        },
      },
    },
  },
  Object {
    "name": Object {
      "props": Object {
        "label": "The name",
        "value": "current value of name",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
    "phone": Object {
      "props": Object {
        "label": "The phone",
        "value": "current value of phone",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
  },
  "i validate forms",
  [Function],
]
`;

exports[`getValuesIfFormValid returns null if not valid 2`] = `
Array [
  Object {
    "obj": "containing errors",
  },
]
`;

exports[`getValuesIfFormValid returns values if form is valid  1`] = `"merged form values"`;

exports[`no submit if validation fails 1`] = `
Array [
  Immutable.Map {
    "fields": Immutable.Map {
      "name": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The name",
        },
        "current": Immutable.Map {
          "value": "current value of name",
        },
      },
      "phone": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The phone",
        },
        "current": Immutable.Map {
          "value": "current value of phone",
        },
      },
    },
  },
  Object {
    "name": Object {
      "props": Object {
        "label": "The name",
        "value": "current value of name",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
    "phone": Object {
      "props": Object {
        "label": "The phone",
        "value": "current value of phone",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
  },
  [MockFunction],
  [Function],
]
`;

exports[`no submit if validation fails 2`] = `
Array [
  Object {
    "obj": "containing errors",
  },
]
`;

exports[`no submit if validation fails 3`] = `
Array [
  Object {
    "payload": Object {
      "obj": "containing errors",
    },
    "type": "validateAll",
  },
]
`;

exports[`options passed to createReducer 1`] = `
Array [
  Object {
    "fields": Array [
      Object {
        "label": "The name",
        "name": "name",
        "type": "text",
      },
      Object {
        "label": "The phone",
        "name": "phone",
        "type": "text",
      },
    ],
    "formTools": Object {
      "current": Object {
        "addField": [Function],
        "addFieldListener": [Function],
        "getFields": [Function],
        "getValues": [Function],
        "removeField": [Function],
        "removeFieldListener": [Function],
        "setValue": [Function],
        "updateFieldDefinition": [Function],
      },
    },
    "initialValues": Immutable.Map {},
    "options": Object {
      "optional": "yes",
    },
  },
]
`;

exports[`removeFieldListener dispatches action 1`] = `
Array [
  Object {
    "fieldName": "formField",
    "payload": "myListener",
    "type": "removeListener",
  },
]
`;

exports[`reset dispatches action with derived initial state 1`] = `
Array [
  Array [
    Object {
      "label": "The name",
      "name": "name",
      "type": "text",
    },
    Object {
      "label": "The phone",
      "name": "phone",
      "type": "text",
    },
  ],
  Immutable.Map {
    "nested": Immutable.Map {
      "name": "old name",
    },
    "id": "id",
  },
  Object {},
  Object {
    "current": Object {
      "addField": [Function],
      "addFieldListener": [Function],
      "getFields": [Function],
      "getValues": [Function],
      "removeField": [Function],
      "removeFieldListener": [Function],
      "setValue": [Function],
      "updateFieldDefinition": [Function],
    },
  },
]
`;

exports[`reset dispatches action with derived initial state 2`] = `
Array [
  Object {
    "payload": "derived-initial-state",
    "type": "reset",
  },
]
`;

exports[`returns array of form props 1`] = `
Array [
  Object {
    "name": Object {
      "id": "name",
      "label": "The name",
      "value": "current value of name",
    },
    "phone": Object {
      "id": "phone",
      "label": "The phone",
      "value": "current value of phone",
    },
  },
  Object {
    "Form": [Function],
    "addField": [Function],
    "addFieldListener": [Function],
    "getAnyTouched": [Function],
    "getIsPristine": [Function],
    "getValues": [Function],
    "getValuesIfFormValid": [Function],
    "id": "",
    "removeField": [Function],
    "removeFieldListener": [Function],
    "reset": [Function],
    "setValue": [Function],
    "submit": [Function],
    "updateFieldDefinition": [Function],
    "validate": [Function],
  },
]
`;

exports[`submit with skipValidation must be a boolean 1`] = `
Array [
  Immutable.Map {
    "fields": Immutable.Map {
      "name": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The name",
        },
        "current": Immutable.Map {
          "value": "current value of name",
        },
      },
      "phone": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The phone",
        },
        "current": Immutable.Map {
          "value": "current value of phone",
        },
      },
    },
  },
  Object {
    "name": Object {
      "props": Object {
        "label": "The name",
        "value": "current value of name",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
    "phone": Object {
      "props": Object {
        "label": "The phone",
        "value": "current value of phone",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
  },
  "i validate forms",
  [Function],
]
`;

exports[`valid form submits values  1`] = `
Array [
  "merged form values",
]
`;

exports[`valid form submits values  2`] = `
Array [
  Immutable.Map {
    "fields": Immutable.Map {
      "nested.name": Immutable.Map {
        "initial": Immutable.Map {
          "type": undefined,
          "label": undefined,
        },
        "current": Immutable.Map {
          "value": "current value of nested.name",
        },
      },
      "phone": Immutable.Map {
        "initial": Immutable.Map {
          "type": undefined,
          "label": undefined,
        },
        "current": Immutable.Map {
          "value": "current value of phone",
        },
      },
    },
  },
  Immutable.Map {
    "nested": Immutable.Map {
      "name": "old name",
    },
    "id": "id",
  },
]
`;

exports[`valid form submits values  3`] = `
Array [
  Object {
    "payload": Object {
      "obj": "containing errors",
    },
    "type": "validateAll",
  },
]
`;

exports[`validate can be an array of validators 1`] = `
Array [
  Immutable.Map {
    "fields": Immutable.Map {
      "name": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The name",
        },
        "current": Immutable.Map {
          "value": "current value of name",
        },
      },
      "phone": Immutable.Map {
        "initial": Immutable.Map {
          "type": "text",
          "label": "The phone",
        },
        "current": Immutable.Map {
          "value": "current value of phone",
        },
      },
    },
  },
  Object {
    "name": Object {
      "props": Object {
        "label": "The name",
        "value": "current value of name",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
    "phone": Object {
      "props": Object {
        "label": "The phone",
        "value": "current value of phone",
      },
      "setValidationResult": [MockFunction],
      "setValue": [MockFunction],
      "validate": [MockFunction],
    },
  },
  Array [
    "i validate forms",
    "as do i",
  ],
  [Function],
]
`;
